src/core/pointcloud/qgseptdecoder.cpp

```diff
@@ -672,4 +672,303 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QByteArray &byteArrayDat
+              break;
+            case QgsPointCloudAttribute::UChar:
+              _storeToStream<unsigned char>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<unsigned char * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::Short:
+              _storeToStream<qint16>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<qint16 * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::UShort:
+              _storeToStream<quint16>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<quint16 * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::Int32:
+              _storeToStream<qint32>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<qint32 * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::UInt32:
+              _storeToStream<quint32>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<quint32 * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::Int64:
+              _storeToStream<qint64>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<qint64 * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::UInt64:
+              _storeToStream<quint64>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<quint64 * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::Float:
+              _storeToStream<float>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<float * >( &buf[requestedAttribute.offset] ) );
+              break;
+            case QgsPointCloudAttribute::Double:
+              _storeToStream<double>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<double * >( &buf[requestedAttribute.offset] ) );
+              break;
+          }
+        }
+        break;
+        case LazAttribute::MissingOrUnknown:
+          // just store 0 for unknown/missing attributes
+          _storeToStream<unsigned short>( dataBuffer, outputOffset, requestedAttribute.type, 0 );
+          break;
+      }
+
+      outputOffset += requestedAttribute.size;
+    }
+
+    // check if point needs to be filtered out
+    if ( filterIsValid )
+    {
+      // we're always evaluating the last written point in the buffer
+      double eval = filterExpression.evaluate( i - skippedPoints );
+      if ( !eval || std::isnan( eval ) )
+      {
+        // if the point is filtered out, rewind the offset so the next point is written over it
+        outputOffset -= requestedPointRecordSize;
+        ++skippedPoints;
+      }
+    }
+  }
+
+  block->setPointCount( pointCount - skippedPoints );
+  return block.release();
+}
```

Comment:
It would be probably good to move any lazperf-related code to a static functions in a new class (e.g. QgsLazperfUtils) which would not be a part of public API. And COPC would not need QgsEptDecoder class at all, it would only use functions from the new one.